src/swap/useSwapQuote.ts

```diff
@@ -20,88 +21,98 @@ const useSwapQuote = () => {
-    setFetchingSwapQuote(false)
-  }, [exchangeRate])
 
   // refreshQuote requests are generated when the swap input amounts are
   // changed, but the quote response / updated exchange rate updates the swap
   // input amounts. this variable prevents duplicated requests in this scenario
   const requestUrlRef = useRef<string>('')
 
-  const refreshQuote = async (
-    fromToken: TokenBalance,
-    toToken: TokenBalance,
-    swapAmount: ParsedSwapAmount,
-    updatedField: Field
-  ) => {
-    setFetchSwapQuoteError(false)
+  const refreshQuote = useAsyncCallback(
+    async (
+      fromToken: TokenBalance,
+      toToken: TokenBalance,
+      swapAmount: ParsedSwapAmount,
+      updatedField: Field
+    ) => {
+      if (!swapAmount[updatedField].gt(0)) {
+        setExchangeRate(null)
+        return null
+      }
 
-    if (!swapAmount[updatedField].gt(0)) {
-      setExchangeRate(null)
-      return
-    }
+      // This only works for tokens with 18 decimals
+      // TODO: make this work for tokens with different decimals
+      const swapAmountInWei = multiplyByWei(swapAmount[updatedField])
+      if (swapAmountInWei.lte(0)) {
+        setExchangeRate(null)
+        return null
+      }
 
-    // This only works for tokens with 18 decimals
-    // TODO: make this work for tokens with different decimals
-    const swapAmountInWei = multiplyByWei(swapAmount[updatedField])
-    if (swapAmountInWei.lte(0)) {
-      setExchangeRate(null)
-      return
-    }
+      const swapAmountParam = updatedField === Field.FROM ? 'sellAmount' : 'buyAmount'
+      const params = {
+        buyToken: toToken.address,
+        sellToken: fromToken.address,
+        [swapAmountParam]: swapAmountInWei.toFixed(0, BigNumber.ROUND_DOWN),
+        userAddress: walletAddress ?? '',
+      }
+      const queryParams = new URLSearchParams({ ...params }).toString()
+      const requestUrl = `${networkConfig.approveSwapUrl}?${queryParams}`
+      if (requestUrl === requestUrlRef.current) {
+        // do nothing if the previous request url is the same as the current
+        return null
+      }
 
-    const swapAmountParam = updatedField === Field.FROM ? 'sellAmount' : 'buyAmount'
-    const params = {
-      buyToken: toToken.address,
-      sellToken: fromToken.address,
-      [swapAmountParam]: swapAmountInWei.toFixed(0, BigNumber.ROUND_DOWN),
-      userAddress: walletAddress ?? '',
-    }
-    const queryParams = new URLSearchParams({ ...params }).toString()
-    const requestUrl = `${networkConfig.approveSwapUrl}?${queryParams}`
-    if (requestUrl === requestUrlRef.current) {
-      // do nothing if the previous request url is the same as the current
-      return
-    }
+      requestUrlRef.current = requestUrl
+      const response = await fetch(requestUrl)
+
+      if (!response.ok) {
+        throw new Error(await response.text())
+      }
 
-    requestUrlRef.current = requestUrl
+      return {
+        response,
+        requestParams: {
```

Comment:
from the `onSuccess` callback i couldn't figure out how to access the function arguments :/ open for suggestions on how better to do this part